pkg/config/cluster_config.go

```diff
@@ -110,28 +114,53 @@ func GetTokenExchangeURLfromStorageSecretStore(config Config, providerType strin
 		return "", utils.Error{Description: utils.WarnFetchingTokenExchangeURL}
 	}
 
-	// If the cluster is not satellite cluster, use PROD or STAGE URLs
-	if os.Getenv("IS_SATELLITE") != constTrue {
-		if !strings.Contains(url, "stage") && !strings.Contains(url, "test") {
-			url = utils.ProdPrivateIAMURL
-		} else {
-			url = utils.StagePrivateIAMURL
+	isProd := isProduction(url)
+
+	// If the cluster is satellite, always use public IAM URL.
+	if isSatellite {
+		if isProd {
+			return utils.ProdPublicIAMURL + tokenExchangePath, nil
 		}
+		return utils.StagePublicIAMURL + tokenExchangePath, nil
 	}
 
-	// Appending the base URL and token exchange path
-	url = url + tokenExchangePath
-
-	return url, nil
+	if isProd {
+		return utils.ProdPrivateIAMURL + tokenExchangePath, nil
+	}
+	return utils.StagePrivateIAMURL + tokenExchangePath, nil
 }
 
 // FrameTokenExchangeURLFromClusterInfo ...
-func FrameTokenExchangeURLFromClusterInfo(cc ClusterConfig, logger *zap.Logger) string {
+func FrameTokenExchangeURLFromClusterInfo(isSatellite bool, cc ClusterConfig, logger *zap.Logger) string {
+
 	if !strings.Contains(cc.MasterURL, stageMasterURLsubstr) {
 		logger.Info("Env-Production")
+		if isSatellite {
+			return (utils.ProdPublicIAMURL + tokenExchangePath)
+		}
 		return (utils.ProdPrivateIAMURL + tokenExchangePath)
 	}
 
 	logger.Info("Env-Stage")
+	if isSatellite {
+		return (utils.StagePublicIAMURL + tokenExchangePath)
+	}
 	return (utils.StagePrivateIAMURL + tokenExchangePath)
 }
+
+// isProduction determines if the env in which a pod is deployed is stage or production
+func isProduction(url string) bool {
+	if !strings.Contains(url, "stage") && !strings.Contains(url, "test") {
+		return true
+	}
+	return false
+}
+
+// IsSatellite checks if the cluster where the pod is currently running is a satellite cluster or not
+func IsSatellite(cc ClusterConfig, logger *zap.Logger) bool {
```

Comment:
Check if configmap cluster-info is editable

Reply:
going forward lets plan to use get cluster type from immutable object